ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -311,20 +307,26 @@ txSubmissionInbound _tracer maxUnacked mpReader mpWriter _version =
             -- Check if having received more txs we can now confirm any (in
             -- strict order in the unacknowledgedTxIds sequence).
             (acknowledgedTxIds, unacknowledgedTxIds') =
-              Seq.spanl (`Map.member` bufferedTxs') (unacknowledgedTxIds st)
+              Seq.spanl (`Map.member` btxsReq) (unacknowledgedTxIds st)
 
             -- If so we can submit the acknowledged txs to our local mempool
-            txsReady = foldr (\txid r -> maybe r (:r) (bufferedTxs' Map.! txid))
+            txsReady = foldr (\txid r -> maybe r (:r) (btxsReq Map.! txid))
                              [] acknowledgedTxIds
 
             -- And remove acknowledged txs from our buffer
-            bufferedTxs'' = foldl' (flip Map.delete)
-                                   bufferedTxs' acknowledgedTxIds
+            btxsAcked = foldl' (flip Map.delete)
+                                   btxsReq acknowledgedTxIds
+
+        _writtenTxids <- mempoolAddTxs txsReady
 
-        _ <- mempoolAddTxs txsReady
+        -- If we are acknowleding transactions that are still in unacknowledgedTxIds'
+        -- we need to re-add them so that we also can acknowledge them again later.
+        -- This will happen incase of duplicate txids within the same window.
+        let live =  filter (`elem` unacknowledgedTxIds') $ toList acknowledgedTxIds
+            btxsReAdded = forceElemsToWHNF $ btxsAcked <> (Map.fromList (zip  live (repeat Nothing)))
```

Comment:
Does this `forceElemsToWHNF` force the thunks? Because the type of `bufferedTxs` is `Map txid (Maybe tx)`, so even forcing a `Maybe tx` to WHNF won't force the `tx` to WHNF (unless it already is).

What we in consensus typically do is replace `Maybe` with a strict variant, often specific to the use case. The constructor names would then also explain what `Nothing` and `Just` mean.

Actually, could you try removing the `forceElemsToWHNF` and replace `<>` with `Map.union`? I think the problem might actually be that `Map.Strict` is the same type as `Map.Lazy` and that the `Semigroup` instance of `Map.Lazy` is being used here.

Reply:
Removing `forceElemsToWHNF` and replacing `<>` with `Map.union` seems to still work. However I seem to be unable to trigger a "thunk invariance failure" at all so I'm not going to address this.

Reply:
I still think that `Map.union` is cleaner than `forceElemsToWHNF` (no need to force thunks if you don't even introduce them).